Public/Src/Cache/ContentStore/Vsts/DedupReadOnlyContentSession.cs

```diff
@@ -155,14 +172,62 @@ public class DedupReadOnlyContentSession : ContentSessionBase
                 return new PinResult($"DedupStore client requires {RequiredHashType}. Cannot take HashType '{contentHash.HashType}'.");
             }
 
-            try
+            var pinResult = CheckPinInMemory(contentHash);
+            if (pinResult.Succeeded)
             {
-                var pinResult = CheckPinInMemory(contentHash);
-                if (pinResult.Succeeded)
-                {
-                    return pinResult;
-                }
+                return pinResult;
+            }
+
+            // Since pinning the whole tree can be an expensive operation, we have optimized how we call it. Depending on the current
+            //  keepUntil of the root node, which is unexpensive to check, the operation will behave differently:
+            //      The pin operation will be ignored if it is greater than ignorePinThreshold, to reduce amount of calls
+            //      The pin operation will be inlined if it is lower than pinInlineThreshold, to make sure that we don't try to use
+            //          content that we pin in the background but has expired before we could complete the pin.
+            //      The pin operation will be done asynchronously and will return success otherwise. Most calls should follow this

```

Comment:
I'm a little bit concerned about this pattern, because if the pin will fail, it won't be possible to handle it.
And as you mentioned in this comment, this is the most common case.

Do we know that this is just fine to ignore the results for such a crucial operation?

Reply:
Discussed in person and was cleared up in the comment